### PR TITLE
Build into build.new and move it live only when it is complete

### PR DESCRIPTION
**Pull request by gounthar**
Earlier today, the build of this repo failed, and the result (which was empty because of the failure) was published.
This linked to a 404.
The issue was quickly solved by correcting the faulty document syntax, but this does not prevent future failures of the same sort.
This is a naive attempt at slightly hardening the process: the repo now builds into `build.new` and moves that into place only on success.

### CODE REVIEW

**Review by stormi — COMMENTED** [review]
@gounthar I would prefer a shorter and personal commit message.

**Comment by stormi** [comment]
> @gounthar I would prefer a shorter and personal commit message.

It's definitetely shorter, though I hope it's not the result of asking the LLM to condense it and make it personal (so many "I" in the sentences now, that's not really what I meant with personal. I just meant: your own words). I doesn't feel like one's own commit message, but at least it's straight to the point. It would be even better, conforming to the established conventions, if it would start with a short description of why the change is necessary, as here the PR starts without context.

File: deploy.sh [hunk 1]
```diff
@@ -1,14 +1,58 @@
...
-PATH="/bin:/usr/bin:/user/local/bin:/var/www/n/bin"
+# Pull master and rebuild the documentation site.
+#
+# The build goes to a scratch directory and is only moved into place once it has
+# succeeded, so a commit that does not build can no longer take the site down.
+#
+# This matters because "docusaurus build" empties its output directory before
+# webpack starts, not after it finishes: see clearPath(outDir) in
+# @docusaurus/core/lib/commands/build/buildLocale.js, which runs in the same
+# Promise.all as the config generation. Building straight into the live
```
Comment (thomas-dkmt) [inline deploy.sh]: I would move 'Therefore' at the start of the sentence, before 'Building', to make the sentence sound a little more natural.

File: deploy.sh [hunk 2]
```diff
@@ -1,14 +1,58 @@
...
+
+PATH="/bin:/usr/bin:/usr/local/bin:/var/www/n/bin"
+
+cd "$(dirname "$0")"
+
+live="build"
+staging="build.new"
+previous="build.prev"
+
+# Keep one previous log. Until now each run overwrote it, so the log of a failed
```
Comment (thomas-dkmt) [inline deploy.sh]: Would _'Keep one past log'_ be OK? When I read 'previous' I think 'previous to what ?' and it sounds a little awkward, but maybe that's just me.

Also, we should add a comma after 'Until now'.

File: deploy.sh [hunk 3]
```diff
@@ -1,14 +1,58 @@
...
-# Generate the doc
-npm run build
-end=`date`
-echo "Pull and build done at $end"
+
+# Generate the doc into staging, never into the directory being served
+rm -rf "$staging"
+npm run build -- --out-dir "$staging"
+
+# Only now is there a complete site to put in place
```
Comment (thomas-dkmt) [inline deploy.sh]: How about _'# The new site is now fully built and ready to go live'_? Sounds more natural to me

**Review by thomas-dkmt — CHANGES_REQUESTED** [review]
I'm not sure how I can test the script myself, but I left a few suggestions to improve the style of your code comments.

File: deploy.sh (hunk 1 above)
Comment (gounthar) [reply deploy.sh]: Done, thanks.

File: deploy.sh (hunk 2 above)
Comment (gounthar) [reply deploy.sh]: Done, thanks.

File: deploy.sh (hunk 3 above)
Comment (gounthar) [reply deploy.sh]: Done, thanks.

**Review by thomas-dkmt — APPROVED** [review]

File: deploy.sh [hunk 4]
```diff
@@ -1,14 +1,58 @@
...
+# succeeded, so a commit that does not build can no longer take the site down.
+#
+# This matters because "docusaurus build" empties its output directory before
+# webpack starts, not after it finishes: see clearPath(outDir) in
+# @docusaurus/core/lib/commands/build/buildLocale.js, which runs in the same
+# Promise.all as the config generation. Therefore, building straight into the
+# live directory leaves nothing to serve for the whole build, and nothing
+# at all if the build then fails.
+
+set -Eeuo pipefail
```
Comment (nathanael-h) [inline deploy.sh:14]: What is the option `-E`?

**Comment by nathanael-h** [comment]
Was the new deployment script tested, with all cases possible (build fails, build success, the different directories are present or not)? 

@thomas-dkmt you can test on any Linux (even WSL on Windows) machine with `./deploy.sh`

File: deploy.sh (hunk 4 above)
Comment (gounthar) [reply deploy.sh:14]: As far as I know, it's the short form of -o errtrace, which is most certainly unnecessary for this script. 🤷 
Shall I remove it?

Comment (nathanael-h) [reply deploy.sh:14]: It might be needed for `trap on_failure ERR` to be called on failure. Please share test results.

**Comment by gounthar** [comment]
@thomas-dkmt was kind enough to spend some time with me this morning to evaluate the changes.
I may have to spend some more time on the edge cases.

**Comment by gounthar** [comment]
I did all the tests I could think of, and they went fine.
I haven't changed anything in my proposed `deploy.sh` after the tests, but I had to edit it before the tests because it starts by checking out master and wiping out all the introduced errors that I needed to test the new version of the script. 🤷 

If this PR gets merged, I'll then create a draft one that would use a GitHub action to build the website on each PR, so we wouldn't merge a PR that produces a broken website (that wouldn't be published thanks to the new "guards" in `deploy.sh`)

**Comment by nathanael-h** [comment]
It's great that tests are passing. Then I see no blocker. I did not get why the PR was set to Draft? 

Also very good idea to build against PR, maybe we should have a manual approval of CI run for people not members of the xcp-ng org (to guard from CI poisoning) and automatic run for members opening PR?

**Comment by gounthar** [comment]
I set it to draft as I hadn't tested all I could, just the few tests I could think of, and I didn't want anyone to have this PR in their "TO REVIEW" list until all is set on my side.
